### PR TITLE
Implement SetChargingProfileRequest and basic database storage/loading

### PR DESCRIPTION
**Pull request by christopher-davis-afs**
## Describe your changes

Implements handling SetChargingProfileRequest from K01

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

Depends on #682

### CODE REVIEW

**Comment by christopher-davis-afs** [comment]
CC: @shankari

File: config/v201/core_migrations/1_up-initial.sql
Comment (marcemmers) [inline config/v201/core_migrations/1_up-initial.sql:90]: Like mentioned in the weekly call, please create a new up (and down) file for these changes.

File: config/v201/core_migrations/1_up-initial.sql
Comment (marcemmers) [inline config/v201/core_migrations/1_up-initial.sql:92]: Looking at the get and clear charging profile functionality it might be beneficial to also store the stacklevel and chargingProfilePurpose as separate values in the database, next to the json. That would allow us to get and delete directly on the database layer for most of those calls.

File: include/ocpp/v201/charge_point.hpp
Comment (marcemmers) [inline include/ocpp/v201/charge_point.hpp:782]: I've mentioned this before but I don't see a good way out of it yet. I don't like making stuff protected just for unit testing. I think we might have to look at a better solution for the library in general to improve testing capabilities. Don't expect you to make any changes based on this comment, if you guys have idea's that's always welcome of course ;)

**Review by marcemmers — CHANGES_REQUESTED** [review]

**Comment by shankari** [comment]
@louisg1337 @the-bay-kay can you pre-review this before I take a look?

File: config/v201/core_migrations/1_up-initial.sql
Comment (christopher-davis-afs) [reply config/v201/core_migrations/1_up-initial.sql:90]: What should be in the `down` file? We don't want to delete profiles when we shut down, right?

Comment (marcemmers) [reply config/v201/core_migrations/1_up-initial.sql:90]: The down file is only used when an older version of libocpp is installed so that we can go back to the exact schema that version expects. If you add a table like you are doing here, the down file should just drop the table.

See https://github.com/EVerest/libocpp/blob/main/doc/database_migrations.md

File: lib/ocpp/v201/charge_point.cpp
Comment (christopher-davis-afs) [reply lib/ocpp/v201/charge_point.cpp:3854]: What do you have in mind?

**Comment by christopher-davis-afs** [comment]
I'm very confused by the tests segfaulting on the last few CI runs. They do not segfault locally.

**Comment by marcemmers** [comment]
> I'm very confused by the tests segfaulting on the last few CI runs. They do not segfault locally.

I am not sure if it's related but I have seen issues with the test databases before. It was the other way around then though, Segfaults locally and not on the server.

In general I think it's better to unit test against in memory databases instead of actual file databases. In this case I would replace `/tmp/ocpp201/cp.db` with an in memory database. Make sure you don't use the same cache as the device model does ;)

File: lib/ocpp/v201/charge_point.cpp
Comment (marcemmers) [inline lib/ocpp/v201/charge_point.cpp:3857]: You can use `for (auto& [evse_id, profile] : evses)` here for increased readability.

File: lib/ocpp/v201/charge_point.cpp
Comment (marcemmers) [reply lib/ocpp/v201/charge_point.cpp:3854]: You could make this function return a bool and then clear the database when you get `false` on the call for example. There are probably more elegant solutions to deal with this but this at least ensures you can continue after.

It's still a big issue if it happens at all since it means we can corrupt the database in some way and should find out what causes that.

**Comment by maaikez** [comment]
> > I'm very confused by the tests segfaulting on the last few CI runs. They do not segfault locally.
> 
> I am not sure if it's related but I have seen issues with the test databases before. It was the other way around then though, Segfaults locally and not on the server.
> 
> In general I think it's better to unit test against in-memory databases instead of actual file databases. In this case I would replace `/tmp/ocpp201/cp.db` with an in memory database. Make sure you don't use the same cache as the device model does ;)

I did make some changes on the database stuff, when it opens I am adding some flags. Can that be the reason for recent crashes?

The in memory database did not work because those flags are missing, but if you use the correct flags, it will work. I did not look into the test to see if they use those flags but if you want to change to an in memory database, you have to check that.

**Comment by maaikez** [comment]
> > > I'm very confused by the tests segfaulting on the last few CI runs. They do not segfault locally.
> > 
> > 
> > I am not sure if it's related but I have seen issues with the test databases before. It was the other way around then though, Segfaults locally and not on the server.
> > In general I think it's better to unit test against in-memory databases instead of actual file databases. In this case I would replace `/tmp/ocpp201/cp.db` with an in memory database. Make sure you don't use the same cache as the device model does ;)
> 
> I did make some changes on the database stuff, when it opens I am adding some flags. Can that be the reason for recent crashes?
> 
> The in memory database did not work because those flags are missing, but if you use the correct flags, it will work. I did not look into the test to see if they use those flags but if you want to change to an in memory database, you have to check that.

Having said, this was when not looking into the logs. I now looked into the logs but I can not really see why the tests fail. They seem to stop, or crash maybe, but it is not really clear to me which tests fail and why.

**Comment by christopher-davis-afs** [comment]
The tests didn't segfault this time for some reason?

File: tests/lib/ocpp/v201/test_database_handler.cpp
Comment (couryrr-afs) [reply tests/lib/ocpp/v201/test_database_handler.cpp]: I was running into a locking issues and this allowed the step to occur again. I updated the call without a loop.

File: include/ocpp/v201/charge_point.hpp
Comment (couryrr-afs) [reply include/ocpp/v201/charge_point.hpp:782]: We agree and once we have things more fully delivered we can take a look at improving on this.

File: config/v201/core_migrations/1_up-initial.sql
Comment (couryrr-afs) [reply config/v201/core_migrations/1_up-initial.sql]: I like this for a long term solution. Could this be part of this [issue](https://github.com/EVerest/libocpp/issues/713)?

File: include/ocpp/v201/charge_point.hpp
Comment (marcemmers) [reply include/ocpp/v201/charge_point.hpp:782]: Last Monday I had a discussion with the people from Pionix. We have some ideas on how to improve the way we can test charge_point and more specifically the message handling. Nobody will pick it up yet due to other priorities but that will help for this issue as well.

File: config/v201/core_migrations/1_up-initial.sql
Comment (marcemmers) [reply config/v201/core_migrations/1_up-initial.sql]: If possible I would prefer for you to add it to this PR already, even if you don't really implement this yet. If we don't do it in this PR we need to add an additional migration file for adding these fields. I can give you a proposal for this if that would help?

Comment (couryrr-afs) [reply config/v201/core_migrations/1_up-initial.sql]: I can get the update done. Do you want the crud portions in the `DatabaseHandler` or just the table updates?

Comment (marcemmers) [reply config/v201/core_migrations/1_up-initial.sql]: The crud portions can be part of the clear and get charging profile use cases I think so just the table updates should be fine.

Comment (couryrr-afs) [reply config/v201/core_migrations/1_up-initial.sql]: This is done.

File: lib/ocpp/v201/charge_point.cpp
Comment (christopher-davis-afs) [reply lib/ocpp/v201/charge_point.cpp]: Added a new method that validates and adds in one step.

File: lib/ocpp/v201/charge_point.cpp
Comment (couryrr-afs) [reply lib/ocpp/v201/charge_point.cpp:3863]: I slightly refactored the try/catch removing two cases. I am not sure about the idea of fully deleting the database; it seems a bit too aggressive. Currently, there are three ways to fail. This is ignoring the idea that the DB is corrupted.

The only way I see for any of these to happen is if the query is broken. We currently have test which should ensure this does not occur. In the case of bad data making it into the DB then failures would be a good indication of a system issue.

Load Failure - A call to get all profiles by the evse is done. If this query fails we could just let the system run and the CSMS could resend the data. The other option would be to not boot.

Add Failure - During boot the system validates profiles and adds them back to the in-memory cache. There is a call to add these to the database, which I find odd, but if this fails it does not matter. The profile is already present.

Delete Failure - This is the only case that would cause divergence in the two states. The in-memory database would not have something the database does. Since all processing is on the in-memory cache nothing critical should fail.

**Review by shankari — CHANGES_REQUESTED** [review]
@christopher-davis-afs @couryrr-afs @wjmp @drmrd I have requested short descriptions/outlines of functionality as part of the commit messages or PRs before, notably in context of https://github.com/EVerest/libocpp/pull/685#issuecomment-2211138420 and in sprint review meetings, but I don't see that in this PR. Do you have an ETA of when you plan to start including them?

I am fine with all the nits being addressed in a cleanup change. Before approval, I would like to understand:
1. the process for tracking the comments and ensuring that they are addressed
2. at least a high-level design of what the final error handling during load will be; even if the implementation waits until later

File: include/ocpp/v201/charge_point.hpp
Comment (marcemmers) [reply include/ocpp/v201/charge_point.hpp:782]: Yes of course. 

The general idea is to split up the handling of messages per functional block. In this case that would be a class to handle K related messages. The charge point will most likely still receive the messages and forward them to the correct handler. All the dependencies that the handlers might need will be injected by the charge point class. 

With this in place it would be possible to test the handling of the K messages by only constructing an object of the handler and all the other dependencies can be mocked. The charge point does not need to be constructed.

The other way around is also possible. You can construct a charge point with mocked message handlers so you are able to test for example the message routing.

How this will work in practice and if this is the final way of doing it is not sure yet. This was just what we discussed briefly.

File: lib/ocpp/v201/smart_charging.cpp
Comment (christopher-davis-afs) [reply lib/ocpp/v201/smart_charging.cpp:80]: `reasonCode` is a specific set of values listed in the OCPP v201 Appendices document, with the accepted values being laid out and what they should be used for. We put the detailed information in `additionalInfo`.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (marcemmers) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:207]: Indeed this was by my request so that all these functions have the same argument order :+1:

File: lib/ocpp/v201/database_handler.cpp
Comment (couryrr-afs) [reply lib/ocpp/v201/database_handler.cpp:762]: Updated name to `get_all_charging_profiles_group_by_evse`. I am batching up my changes so it is not present yet.

File: include/ocpp/v201/charge_point.hpp
Comment (couryrr-afs) [reply include/ocpp/v201/charge_point.hpp:782]: Would it be useful to add this to an issue within the `EVerest/libocpp` repo so the community can better track this? I do not think we are expected to take any action within this PR.

Comment (shankari) [reply include/ocpp/v201/charge_point.hpp:782]: Just to clarify, I am not asking for any changes other than:
- the process for tracking the comments and ensuring that they are addressed
- at least a high-level design of what the final error handling during load will be; even if the implementation waits until later

File: doc/ocpp_201_status.md
Comment (christopher-davis-afs) [reply doc/ocpp_201_status.md:1248]: [This commit explains it](https://github.com/EVerest/libocpp/pull/711/commits/f5b979a58b9d2a642c5536131004821894deacba); We had some FRs that we realized were already covered, but not marked in the documentation.

File: config/v201/core_migrations/5_up-charging_profiles_db.sql
Comment (christopher-davis-afs) [reply config/v201/core_migrations/5_up-charging_profiles_db.sql:1]: This file is a migration file that spins up the `CHARGING_PROFILES` table in the core SQLite database. We don't directly modify the existing DB files for backwards compatibility reasons. See https://github.com/EVerest/libocpp/pull/711#discussion_r1689984569 for more information.

If you're asking why we have both a DB and in-memory storage, that is something we carry over from the 1.6 implementation. We hope to address that soon, per https://github.com/EVerest/libocpp/issues/713

File: include/ocpp/v201/charge_point.hpp
Comment (christopher-davis-afs) [reply include/ocpp/v201/charge_point.hpp:524]: Should this be in the PR description? Or did you mean somewhere else?

Comment (shankari) [reply include/ocpp/v201/charge_point.hpp:524]: Ideally, this would be in:
- the related commit messages *and*
- summarized in the PR description

An example from my NREL project: https://github.com/e-mission/e-mission-phone/pull/1138

File: lib/ocpp/v201/smart_charging.cpp
Comment (shankari) [reply lib/ocpp/v201/smart_charging.cpp:80]: For anybody coming to this later, the appendices is a separate document in the zip file. As of 2024, it is named `OCPP-2.0.1_edition3_part2_appendices_v14.pdf`

**Comment by shankari** [comment]
@couryrr-afs just to clarify, one of the two items I had asked for before approval was:

> at least a high-level design of what the final error handling during load will be; even if the implementation waits until later

Here's the comment around error handling https://github.com/EVerest/libocpp/pull/711#discussion_r1690006967
I don't see a response to it, either in this PR or in the non-blocking comments PR. The "non-blocking PR" also does not include the nits from this PR, only from #611 

If you did add the response elsewhere and I missed it, can you add a link to it here?

Again, I am looking for two high level topics before I approve:
- Tracking the non-blocking comments so they don't get lost
- A high-level description of the design for the final error handling (https://github.com/EVerest/libocpp/pull/711#discussion_r1690006967) so that we can write it down before we forget it

File: lib/ocpp/v201/profile_utils.cpp
Comment (couryrr-afs) [reply lib/ocpp/v201/profile_utils.cpp:10]: The subset of fields was enough for use to check equality in unit tests. We are working to move this to an existing `comparators` file within the testing directory. We can have a more in-depth look at what a proper production ready equality check will be at a later date.

File: lib/ocpp/v201/charge_point.cpp
Comment (couryrr-afs) [reply lib/ocpp/v201/charge_point.cpp:3864]: The comment on database corruption was a bit misleading. The system would have to handle this at a higher level. Specifically, within the `ChargePoint` constructor. Currently, it appears that if an error is throw the system would not boot which is the correct behavior.

Overall, for this method the goal would be [this issue](https://github.com/EVerest/libocpp/issues/713). In this case we would not have to worry about loading any data back into memory and the bulk of this would be fairly minimized. I would highly suggest community buy in on this.

I can add a diagram to the mentioned issue but the flow would be:

1. Boot the system 
  - If a connection cannot be made to the DB, in the constructor, do not boot.
2. Load `ChargingProfiles` for validation
  - `ChargingProfiles` can be invalid if, for example, an EVSE was removed.
  - If the `ChargingProfiles` cannot be loaded do not boot.
3. Delete invalid `ChargingProfiles`
  - If a delete cannot be completed do not boot

In any "do not boot" case we would need to inform the CSMS of the issue.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:1134]: We are setting up an expected bad condition (a TxProfile with a missing transaction id), and expecting `validate_and_add_profile()` to reject the profile. The tests for the response code simply verify that we're rejecting for the expected reasoning. We also test that the profile is not added. This is the behavior of a rejection. Other conditions are covered in our tests for the individual validate and add methods.

Do you have a name that you would prefer?

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (couryrr-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:693]: When going over the requirements for FR.45 again, we realized that we misinterpreted the check that it was asking for. This change is a correction - we compare the number of phases in the profile and reject it if it's greater than what the charging station can supply (AKA ChargingStatisonSupplyPhases). 

Technically 0 is not realistic, but we want to have invalid values in any case. As long as the amount of `SupplyPhases` is less than the given `numberPhases`, we're testing this use case.
 
We don't think a comment is necessary for correcting behavior that was incorrect to begin with.

Comment (shankari) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:693]: Just to clarify for the future: this would be a good comment to put into the commit message that changed the check and to flag in the PR description. Again, the git history (`annotate` in the CLI and `blame` in the UI) will flag that this was changed in a different commit than the rest of the tests. Future community members may want to see why.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:1134]: This is also a non-blocking change; please add it to the list of issues to be tracked and fixed later.

I understand that this is setting up an expected bad condition and testing for rejection. However, there are also other tests which pass in expected bad conditions and test for rejection. My understanding is that this test is one level up from that in that it doesn't just test the failure but that the failure is reported correctly. So a better name would be `K01_ValidateAndAdd_RejectsInvalidProfilesWithReasonCode` or similar.

File: lib/ocpp/v201/charge_point.cpp
Comment (shankari) [reply lib/ocpp/v201/charge_point.cpp:3864]: Agree that we need community buy-in, just wanted to make sure that the current thoughts were being tracked in the longer term discussion.

**Comment by shankari** [comment]
Still not seeing the non-blocking issues from this PR in https://github.com/US-JOET/base-camp/issues/173

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (couryrr-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:693]: I will try to ensure the individual commit messages are a bit more documenting going forward. It is valuable information to include.

**Comment by couryrr-afs** [comment]
> @couryrr-afs just to clarify, one of the two items I had asked for before approval was:
> 
> > at least a high-level design of what the final error handling during load will be; even if the implementation waits until later
> 
> Here's the comment around error handling [#711 (comment)](https://github.com/EVerest/libocpp/pull/711#discussion_r1690006967) I don't see a response to it, either in this PR or in the non-blocking comments PR. The "non-blocking PR" also does not include the nits from this PR, only from #611
> 
> If you did add the response elsewhere and I missed it, can you add a link to it here?
> 
> Again, I am looking for two high level topics before I approve:
> 
>     * Tracking the non-blocking comments so they don't get lost
> 
>     * A high-level description of the design for the final error handling ([Implement SetChargingProfileRequest and basic database storage/loading #711 (comment)](https://github.com/EVerest/libocpp/pull/711#discussion_r1690006967)) so that we can write it down before we forget it

@shankari the tracking ticket on the AFS side has been updated with what I believe to be a full list. 

For the conversation around the database I asked for it to be discussed in the WG, 8/7/2024, and will update the issue referenced with those details. Outside of the comments here is there anything else that is needed for this topic?

**Comment by shankari** [comment]
I spot checked the tracking ticket and it has a couple of issues from here. Giving you the benefit of the doubt that all the pending changes have been tracked. Approving this now.

**Review by shankari — APPROVED** [review]

**Comment by shankari** [comment]
@marcemmers @hikinggrass @Pietfried I haven't seen any requests from the community to fix the static code analysis, so I didn't hold up my approval for it. However, for the projects that I maintain, I generally don't like to regress on the automated checks. Once you do that once, then the check is essentially useless until it is fixed, because the existing errors will mask any new ones. I look forward to seeing how you respond to this as part of your approval....

File: include/ocpp/v201/charge_point.hpp
Comment (marcemmers) [inline include/ocpp/v201/charge_point.hpp:70]: This one is now included twice ;)

File: lib/ocpp/v201/charge_point.cpp
Comment (marcemmers) [reply lib/ocpp/v201/charge_point.cpp:3864]: Currently we don't check the entire database at boot for corruption. This might be a nice addition for the future but I would certainly exclude that from the change of removing the runtime copy of the data and only using the database for storage.

In any case I would also argue that if a delete is not possible we should continue instead of not booting, and still somehow report on this to the CSMS indeed. If we do not boot libocpp we are not even able to contact the CSMS since that is our way of contact.

For now I will resolve this conversation since most of these issues will be solved by #713

**Review by marcemmers — APPROVED** [review]
Just one small comment, otherwise looks good to me :+1:

**Comment by marcemmers** [comment]
@shankari Thanks for bringing it up, I agree we don't always look that well at the static code analysis. Unfortunately this particular tool is flagging false positives sometimes so I have grown a little complacent with it.

For this ticket it is complaining about SQL issues and duplication in test code so for me there are no blockers there. Might be a good idea to disable some of the warnings we don't look at anyway @hikinggrass @Pietfried

**Comment by marcemmers** [comment]
Can this be merged @christopher-davis-afs ? Let me know and I'll press the button.

**Comment by christopher-davis-afs** [comment]
Yes, this is ready :)

**Pull request merged**
